website/docs/sources/discord/index.md

```diff
@@ -255,35 +255,44 @@ if not connection:
 access_token = connection.access_token
 
 guild_member_info = requests.get(
-    GUILD_API_URL.format(guild_id=GUILD_ID),
-    headers={"Authorization": "Bearer " + access_token},
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": "Bearer " + access_token
+  },
 ).json()
 
 # Ensure user is a member of the guild
 if "code" in guild_member_info:
-    if guild_member_info["code"] == 10004:
+    if guild_member_info['code'] == 10004:
         ak_message("User is not a member of the guild")
     else:
-        ak_create_event(
-            "discord_error", source=context["source"], code=guild_member_info["code"]
-        )
+        ak_create_event("discord_error", source=context['source'], code=guild_member_info['code'])
         ak_message("Discord API error, try again later.")
     return False
 
-# Get the user's current roles and remove all roles we want to remap
-new_groups = [
-    role for role in request.user.ak_groups.all() if role not in MAPPED_ROLES.values()
-]
+# Get all discord_groups
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Get all user groups except discord_groups
+user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list('pk', flat=True))
 
-# Add back mapped roles which the user has in the guild
-for role_id in MAPPED_ROLES:
-    if role_id in guild_member_info["roles"]:
-        new_groups.append(MAPPED_ROLES[role_id])
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
+
+# Combine user_groups and matching_roles
+combined_groups = user_groups.union(matching_roles)
 
 # Update user's groups
-request.user.ak_groups.set(new_groups)
+request.user.ak_groups.set(combined_groups)
 request.user.save()
```

Comment:
Saving the user here should not be needed

Reply:
```suggestion
```

Reply:
You are right.